man/systemd-resolved.service.xml

```diff
@@ -263,6 +263,21 @@
     <filename>/run/systemd/resolve/resolv.conf</filename> or lists 127.0.0.53 as DNS server.</para>
   </refsect1>
 
+  <refsect1>
+    <title>resolved-access.d files</title>
+
+    <para>The DNS requests via bus API can be filtered, based on the name of the calling unit, MAC label and
+    domain requested. The filtering parameters are read from
+    <filename>/etc/resolved-access.d/*.conf</filename>, <filename>/usr/lib/resolved-access.d/*.conf</filename> and <filename>/run/resolved-access.d/*.conf</filename>. The files contain the fields like this example:
+    <programlisting># Unit                          MAC Context                     Domain Pattern          Verdict
+systemd-timesyncd.service       system_u:system_r:ntpd_t:s0     *.debian.pool.ntp.org   allow
+systemd-timesyncd.service       system_u:system_r:ntpd_t:s0     *                       deny
+session-*.scope                 user_u:user_r:mozilla_t:s0      *                       allow
```

Comment:
Hmmm. I'm not convinced dns filtering is that useful. Can you describe the intended use cases a bit more?

Why bother with the context? The MAC system can deny access to resolved to the process on its own, I don't see why we should duplicate this capability in resolved. Wouldn't filtering by service be good enough? And if it's by service, wouldn't it be better to make this part of the unit definition? Keeping a centralized db of permissions is repeating the mistakes of existing MAC systems.

Reply:
The idea and purpose is described more in #17053. Here are some actual examples which I use:
1. Allow system administrator and package manager only access distribution package pool and bug reports to update packages:
```
session-*.scope root:sysadm_r:sysadm_t:s0-s0:c0.c1023   *.debian.org    allow
session-*.scope root:sysadm_r:apt_t:s0-s0:c0.c1023      *.debian.org    allow
```
2. Allow systemd-timesyncd only access configured NTP services: 
```
systemd-timesyncd.service       system_u:system_r:ntpd_t:s0     *.debian.pool.ntp.org   allow
```
3. Allow browser and SSH to access anything:
```
session-*.scope user_u:user_r:mozilla_t:s0      *       allow
session-*.scope user_u:user_r:ssh_t:s0          *       allow
```
4. Allow email client to only access configured service provider:
```
session-*.scope user_u:user_r:thunderbird_t:s0  imap.gmail.com allow
```
5. Deny rest:
```
*       *               *       deny
```

> Why bother with the context?
> Wouldn't filtering by service be good enough?

The context is actually important to distinguish user's different application domains, since from service point of view they are just part of `session-*.scope`. We could look at for example `/proc/$SOME_PID/exe` but it can be manipulated while manipulating a MAC context would probably be harder.

Also filtering by UID could be useful.

> The MAC system can deny access to resolved to the process on its own, I don't see why we should duplicate this capability in resolved.

No MAC system has any means to control access individual DNS domains. It may be possible to deny access to DNS entirely with a MAC or something else but while this feature can also do that, it also allows much more fine grained (and actually useful) control.

It might be doable to implement something similar with SELinux networking support and maybe making `resolved` a SELinux object manager. This would mean that `resolved` would tag IP address ranges with peer labels (maybe with NetLabel unlabeled protocol) and node labels (like `semanage node`). Then for example timesyncd would be given TE rules to allow "peer recv" from `timesyncd_peers_t` and  "node sendto" to `timesyncd_nodes_t`. SELinux is not great at deny-listing anything, so blocking something would mean enumerating all allowed peers and nodes. Overlapping ranges of addresses between services could be a problem etc.

Other MAC systems have even worse networking support.
- TOMOYO Linux can allow access to certain IP addresses (`network inet dgram send 0.0.0.0-255.255.255.255 123`)
 but there's no easy way to manage a part of rules since all rules are in one file.
- Smack seems to be able to filter IP addresses much like TOMOYO. I don't know much about it.
- I think AppArmor doesn't know about networks.

This could be implemented globally for a system with a firewall (for example by using IPSet lists) but Linux firewalls sadly are not very good at giving different access to different applications. The best I know of is to combine firewalls with SELinux packet labels but this does not match the need 100%. Part 2 of #17053 is to couple this DNS filtering feature to managing firewalls (for example, when an allowed domain is resolved, the IP is appended to a per-service IPSet). For user applications, Firejail can be a better tool since the application context is more precisely known and it can easily install a firewall in the network namespace and manage it afterwards.

There are of course filtering DNS proxies (for example DNSCrypt-proxy), but they also don't know about systemd services, MAC contexts etc. so the rules apply to the whole system.

> Wouldn't it be better to make this part of the unit definition? Keeping a centralized db of permissions is repeating the mistakes of existing MAC systems.

Maybe, for example per-service directives `DNSAllowedDomains=` and `DNSDeniedDomains=` could work:

systemd-timesyncd.service.d/dns.conf:
```
[Service]
DNSAllowedDomains=mac_context=system_u:system_r:ntpd_t:s0 domain_pattern=*.debian.pool.ntp.org
```
session.d/dns.conf:
```
[Session]
DNSAllowedDomains=mac_context=user_u:user_r:mozilla_t:s0 domain_pattern=*
DNSAllowedDomains=mac_context=user_u:user_r:ssh_t:s0 domain_pattern=*
```
service.d/dns.conf (deny by default):
```
[Service]
DNSDeniedDomains=mac_context=* domain_pattern=*
```

Then resolved would have to get this information from PID1. Is that possible?

This design tries to avoid the centralization by introducing the resolved-access.d directory, then one file (or several) could be added for each application.

Reply:
Correction: AppArmor network rules also use [IP addresses](https://gitlab.com/apparmor/apparmor/-/wikis/AppArmor_Core_Policy_Reference#address-expr) and there seems to be some connection to firewalls.

Reply:
Another point why this must be implemented specifically in `systemd-resolved` instead of DNSCrypt-proxy or unbound etc. is that the other resolvers provide their services to clients with UDP (or TCP for DoH) while resolved can use `nss-resolve` which uses Unix sockets. With UDP and TCP, it's not possible to identify the calling process, even with SELinux (except in very very specific cases). With Unix sockets, we can identify PID, UID, the service etc. Especially the SELinux label of the calling process is determined with SO_PEERSEC, which should not be forgeable. This is impossible with other resolvers.